framework/framework.go

```diff
@@ -142,3 +142,60 @@ func (f *Framework) ProcessDelete(obj interface{}, resources []Resource) error {
 
 	return nil
 }
+
+// ProcessUpdate is a drop-in for an informer's UpdateFunc. It receives the
+// custom object observed during TPR watches and anything that implements
+// Resource. ProcessUpdate takes care about all necessary reconciliation logic
+// for update events. For complex resources this means state has to be created,
+// deleted and updated eventually, in this order.
+//
+//     func updateFunc(oldObj, newObj interface{}) {
+//         err := f.ProcessUpdate(oldObj, newObj, resources)
+//         if err != nil {
+//             // error handling here
+//         }
+//     }
+//
+//     newResourceEventHandler := &cache.ResourceEventHandlerFuncs{
+//         UpdateFunc:    updateFunc,
+//     }
+//
+func (f *Framework) ProcessUpdate(oldObj, newObj interface{}, resources []Resource) error {
```

Comment:
> As I see it an operator needs two pieces of information to be able to reconcile. One is the current state and the other is the desired state.

Agreed.

> The current state has to be looked up using the "old" custom object

I think current state should be retrieved by looking for resources that are already created. For example if operator has to create a deployment it should check if a deployment with a given name already exists. If operator should create an etcd entry it should check if the etcd entry already exists, and so on.

> ,since this is the piece of information that lead to the current state in former iterations. 

Having "old" TPO doesn't mean it was successfully processed. So it isn't information about current state. It's information about desired previous state.

> This can only not work in case the old custom object is empty or wrong

This can be wrong when it failed to process, or the operator pod was killed during processing, or someone deleted one of resources managed by the operator intentionally (for example to force recreating them during next reconciliation).

> Can you elaborate on a specific problem that you see with this approach?

I'll repeat myself: Having "old" TPO doesn't mean it was successfully processed. So it isn't information about current state. It's information about desired previous state.

I think we can't rely on TPO definition as a current state. This can be different in many cases.

Dealing with previous object instance makes operators unnecessary complicated, as they have to be able to process `OnAdd` event with partially created resources anyway.